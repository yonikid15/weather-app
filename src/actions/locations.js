import uuid from 'uuid';
import { countries } from 'country-data';

// ADD_LOCATION
export const addLocation = ( location ) => ({
  type: 'ADD_LOCATION',
  location
});

// ERROR_ADD_LOCATION
export const errorAddLocation = ( message ) => ({
  type: 'ERROR_ADD_LOCATION',
  message
});

export const startAddLocation = ( location = {} ) => {
  const { 
    city, 
    country 
  } = location;

  return ( dispatch ) => {
    return fetchData( 
      location, 
      dispatch, 
      addLocation, 
      errorAddLocation 
    );
  };
};

// REMOVE_LOCATION
export const removeLocation = ( id ) => ({
  type: 'REMOVE_LOCATION',
  id
});

export const startRemoveLocation = ( id ) => {
  return (dispatch) => {
    dispatch( removeLocation( id ) );
  }
};

// REMOVE_ALL_LOCATIONS
export const removeAllLocations = () => ({
  type: 'REMOVE_ALL_LOCATIONS',
});

// ACTIVATE_LOCATION
export const activateLocation = ( id ) => ({
  type: 'ACTIVATE_LOCATION',
  id
});

export const startActivateLocation = ( id ) => {
  return (dispatch) => {
    dispatch( activateLocation( id ) );
  }
};

// REFRESH_LOCATION
export const refreshLocation = ( location ) => ({
  type: 'REFRESH_LOCATION',
  location
});

export const errorRefreshLocation = ( message ) => ({
  type: 'ERROR_REFRESH_LOCATION',
  message
});

export const startRefreshLocation = ( location = {} ) => {
  return ( dispatch ) => {
    return fetchData( 
      location, 
      dispatch, 
      refreshLocation,
      errorRefreshLocation 
    );
  };
};

const fetchData = (location, dispatch, success, error) => {
  const { 
    city, 
    country 
  } = location;

  return fetch(
    `http://api.openweathermap.org/data/2.5/weather?q=${city},${country}&appid=${process.env.OPEN_MAP_WEATHER_API_KEY}`
  ).then( res =>  res.json() )
  .then( data => {
      const locationData = {
          id: data.id,
          temperature: data.main.temp,
          city: data.name,
          country: data.sys.country,
          humidity: data.main.humidity,
          condition: data.weather[0].description,
          active: true
      }
      dispatch( success( locationData ) );
      dispatch( activateLocation( locationData.id ) );
  })
  .catch( err => {
      dispatch( error( err.message ) );
  });
}