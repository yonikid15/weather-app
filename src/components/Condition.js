import React from 'react';

const Condition = ({ condition }) => {
  return (
    <div>
      <h3>
          <span>Condition: </span>
          <span>{condition}</span>
      </h3>
    </div>
  );
};

export default Condition;