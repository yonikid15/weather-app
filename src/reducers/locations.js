// Locations Reducer
const locationsReducerDefaultState = [

];

export default ( state = locationsReducerDefaultState, action ) => {
  switch ( action.type ) {
    case 'ADD_LOCATION':
      return [
        ...state,
        action.location
      ];
    case 'ERROR_ADD_LOCATION': 
      return state;
    case 'REMOVE_LOCATION':
      return state.filter( location => location.id !== action.id );
    case 'REMOVE_ALL_LOCATIONS':
      return [];
    case 'ACTIVATE_LOCATION':
      return state.map( location => {
        if( location.active === true ) {
          location.active = false;
        }

        if( location.id === action.id ) {
          location.active = true;
        }

        return location;
      });
    case 'REFRESH_LOCATION':
      return state.map( location => {
        if( location.id === action.location.id ) {
          return {
            ...action.location
          };
        } else {
          return location;
        }
      });
    case 'ERROR_REFRESH_LOCATION': 
    return state;
    default:
      return state;
  };
};