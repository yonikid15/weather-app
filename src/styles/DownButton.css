.downBtn {
  grid-row: 3;
  background-color: #121519;
  border: 1px solid #607D8B;
  position: fixed;
  width: 50vw;
  height: 50vw;
  left: 30%;
  border-radius: 50%;
  top: 100%;
  transform: translateY(-6vw);
  padding-bottom: 44%;

  &__icon {
    color: #607D8B;
    font-size: 40px;
  }
}

.downBtn:hover {
  background-color: rgba(96, 125, 139, 0.1);
}

.downBtn:focus {
  outline: 0;
  border-color: #00ccff;
  box-shadow: 0 0 0 1px #00ccff;
}