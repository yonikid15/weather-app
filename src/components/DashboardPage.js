import React from 'react';
import { connect } from 'react-redux';
import { startAddLocation, startActivateLocation } from '../actions/locations';
import LocationList from './LocationList';
import LocationForm from './LocationForm';
import Weather from './Weather';
import { shake } from '../tools/animations';
import styles from '../styles/DashboardPage.css';

export const DashboardPage = ( { startAddLocation, startActivateLocation } ) => {
    
    const onSubmit = ( location, ifExists ) => {
        if( !ifExists ) {
            startAddLocation( location );
        } else {
            startActivateLocation( ifExists.id );
            shake();
        }
    }

    return (
        <section className={ styles.layout }>
            {/* Side Content */}
            <div id="side-bar" className={styles.sideBar}>
                <LocationForm onSubmit={onSubmit} />
                <LocationList />
            </div>
            {/* Main Content */}
            <Weather />
        </section>
    );
};

const mapStateToProps = ( state ) => ({
    locations: state.locations
});

const mapDispatchToProps = ( dispatch ) => ({
    startAddLocation: ( location ) => dispatch( startAddLocation( location ) ),
    startActivateLocation: ( id ) => dispatch( startActivateLocation( id ) )
})

export default connect( mapStateToProps, mapDispatchToProps )( DashboardPage );