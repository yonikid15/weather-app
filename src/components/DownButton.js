import React from 'react';
import { connect } from 'react-redux';
import { startActivateLocation } from '../actions/locations';
import { startSlideAnimation } from '../tools/animations';
import styles from '../styles/DownButton.css';

export const DownButton = ( { locations, startActivateLocation } ) => {

  const activate = () => {
    let found = false;

    locations.map( ( location, index ) => {
      if( location.active && locations[ index + 1 ] && !found ) {
        startSlideAnimation({ 
          locations, 
          active: false, 
          id: locations[ index + 1 ].id
        });
        startActivateLocation( locations[ index + 1 ].id );
        found = true;
      }
    })  
  };

  return (
    <button onClick={ activate } className={ styles.downBtn }>
      <i className={`fas fa-caret-down ${ styles.downBtn__icon}`}></i>
    </button>
  );
};

const mapStateToProps = ( state ) => ({
  locations: state.locations
});

const mapDispatchToProps = ( dispatch ) => ({
  startActivateLocation: ( id ) => dispatch( startActivateLocation( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( DownButton );