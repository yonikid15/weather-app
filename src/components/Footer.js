import React from 'react';
import styles from '../styles/Footer.css';

const Footer = () => {
  return (
    <footer className={ styles.footer }>
      {/* Logo */}
      <div>
        yonikid
      </div>
      {/* Link to GitHub */}
      <div>
        view source
      </div>
    </footer>
  );
};

export default Footer;