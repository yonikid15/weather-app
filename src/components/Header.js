import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import styles from '../styles/Header.css';

const Header = () => (
  <header className={styles.header}>
    <div className={styles.header__content}>
      <h1 className={styles.header__title}><Link to="/">Weather Finder</Link></h1>
      <span>{ moment().format('MMMM Do, YYYY') }</span>
    </div>
  </header>
);

export default Header;