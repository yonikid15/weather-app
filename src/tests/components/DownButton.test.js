import React from 'react';
import { shallow } from 'enzyme';
import { DownButton } from '../../components/DownButton';
import locations from '../fixtures/locations';

test('should render DownButton Correctly', () => {
  const wrapper = shallow( <DownButton />);
  expect( wrapper ).toMatchSnapshot();
});

test('should call startActivateLocation on button click', () => {
  document.body.innerHTML = '<div id="activeItem"></div>';
  const onStartActivateLocationSpy = jest.fn();
  const wrapper = shallow( <DownButton locations={locations} startActivateLocation={ onStartActivateLocationSpy }/> );
  wrapper.find( 'button' ).simulate( 'click' );
  expect( onStartActivateLocationSpy ).toHaveBeenCalled();
});

test('should not call startActivateLocation, when next is empty', () => {
  const onStartActivateLocationSpy = jest.fn();
  const wrapper = shallow( <DownButton locations={[]} startActivateLocation={ onStartActivateLocationSpy }/> );
  wrapper.find( 'button' ).simulate( 'click' );
  expect( onStartActivateLocationSpy ).not.toHaveBeenCalled();
});