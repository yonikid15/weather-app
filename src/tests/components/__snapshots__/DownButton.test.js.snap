// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DownButton Correctly 1`] = `
<button
  className="downBtn"
  onClick={[Function]}
>
  <i
    className="fas fa-caret-down downBtn__icon"
  />
</button>
`;
